official/nlp/bert/electra_model.py

```diff
@@ -0,0 +1,323 @@
+    masked_lm_accuracy = numerator / denominator
+
+    discrim_accuracy = tf.keras.metrics.binary_accuracy(tf.cast(discrim_label, tf.float32), discrim_output)
+    self.add_metric(discrim_accuracy, name='discrinator_accuracy', aggregation='mean')
+    self.add_metric(
+        masked_lm_accuracy, name='masked_lm_accuracy', aggregation='mean')
+
+    self.add_metric(lm_example_loss, name='lm_example_loss', aggregation='mean')
+
+    self.add_metric(tot_loss, name='total_loss', aggregation='mean')
+    self.add_metric(discrim_loss, name='discriminator_loss', aggregation='mean')
+
+
+  def call(self, lm_output, lm_label_ids, lm_label_weights,
+           discrim_output, discrim_labels,input_mask):
+    """Implements call() for the layer."""
+    weights = tf.cast(lm_label_weights, tf.float32)
+    lm_output = tf.cast(lm_output, tf.float32)
+    discrim_output = tf.cast(discrim_output, tf.float32)
+    mask_label_loss = losses.weighted_sparse_categorical_crossentropy_loss(
+        labels=lm_label_ids, predictions=lm_output, weights=weights)
+    discrim_ind_loss = tf.nn.sigmoid_cross_entropy_with_logits(logits = discrim_output, labels=tf.cast(discrim_labels,tf.float32))
+    discrim_loss = tf.reduce_sum(discrim_ind_loss)
+    loss = mask_label_loss+self.config["discrim_rate"]*discrim_loss
+
+
+    self._add_metrics(lm_output, lm_label_ids, lm_label_weights,
+                      mask_label_loss, discrim_output, discrim_labels, discrim_loss, loss)
+    return loss
+
+
+@gin.configurable
+def get_transformer_encoder(bert_config,
+                            sequence_length,
+                            transformer_encoder_cls=None):
+  """Gets a 'TransformerEncoder' object.
+
+  Args:
+    bert_config: A 'modeling.BertConfig' or 'modeling.AlbertConfig' object.
+    sequence_length: Maximum sequence length of the training data.
+    transformer_encoder_cls: A EncoderScaffold class. If it is None, uses the
+      default BERT encoder implementation.
+
+  Returns:
+    A networks.TransformerEncoder object.
+  """
+  if transformer_encoder_cls is not None:
+    # TODO(hongkuny): evaluate if it is better to put cfg definition in gin.
+    embedding_cfg = dict(
+        vocab_size=bert_config.vocab_size,
+        type_vocab_size=bert_config.type_vocab_size,
+        hidden_size=bert_config.hidden_size,
+        seq_length=sequence_length,
+        max_seq_length=bert_config.max_position_embeddings,
+        initializer=tf.keras.initializers.TruncatedNormal(
+            stddev=bert_config.initializer_range),
+        dropout_rate=bert_config.hidden_dropout_prob,
+    )
+    hidden_cfg = dict(
+        num_attention_heads=bert_config.num_attention_heads,
+        intermediate_size=bert_config.intermediate_size,
+        intermediate_activation=tf_utils.get_activation(bert_config.hidden_act),
+        dropout_rate=bert_config.hidden_dropout_prob,
+        attention_dropout_rate=bert_config.attention_probs_dropout_prob,
```

Comment:
Please make sure to sync up with latest updates in bert_models.py. For example in bert_models.py there are minor fixes:
kernel_initializer=tf.keras.initializers.TruncatedNormal(
            stddev=bert_config.initializer_range),

Reply:
Should I add the corresponding changes in electra_models?

Reply:
Yes, please.

Reply:
The bert pretrainer now takes different arguments than my electra pretainer, should I change the electra pretrainer so that it is consistent with the bert pretrainer?